File: kotlin/src/main/kotlin/com/github/copilot/JsonRpcClient.kt
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot

import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import java.io.*
import java.nio.charset.StandardCharsets
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * JSON-RPC error response.
 */
class JsonRpcException(
    val code: Int,
    override val message: String,
    val data: JsonElement? = null
) : Exception("JSON-RPC Error $code: $message")

/**
 * Handler for incoming JSON-RPC requests from the server.
 */
typealias RequestHandler = suspend (params: JsonObject) -> JsonObject

/**
 * Minimal JSON-RPC 2.0 client for stdio/TCP transport.
 *
 * Uses a background coroutine for reading and provides a suspend-based interface.
 * Messages are framed with Content-Length headers per the LSP base protocol.
 */
class JsonRpcClient(
    private val inputStream: InputStream,
    private val outputStream: OutputStream,
    private val scope: CoroutineScope
) {
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
    }

    private val pendingRequests = ConcurrentHashMap<String, CompletableDeferred<JsonElement>>()
    private val requestHandlers = ConcurrentHashMap<String, RequestHandler>()

    @Volatile
    private var notificationHandler: ((method: String, params: JsonObject) -> Unit)? = null

    @Volatile
    private var running = false

    private var readJob: Job? = null
    private val writeLock = Any()

    /**
     * Start listening for messages in a background coroutine.
     */
    fun start() {
        if (running) return
        running = true
        readJob = scope.launch(Dispatchers.IO) {
            readLoop()
        }
    }

    /**
     * Stop listening and complete all pending requests with errors.
     */
    fun stop() {
        running = false
        readJob?.cancel()
        readJob = null

        // Complete all pending requests with error
        for ((_, deferred) in pendingRequests) {
            deferred.completeExceptionally(IOException("Client stopped"))
        }
        pendingRequests.clear()
    }

    /**
     * Set handler for incoming notifications from the server.
     */
    fun setNotificationHandler(handler: (method: String, params: JsonObject) -> Unit) {
        this.notificationHandler = handler
    }

    /**
     * Set handler for incoming requests from the server.
     */
    fun setRequestHandler(method: String, handler: RequestHandler?) {
        if (handler == null) {
            requestHandlers.remove(method)
        } else {
            requestHandlers[method] = handler
        }
    }

    /**
     * Send a JSON-RPC request and wait for the response.
     *
     * @param method The RPC method name
     * @param params Optional parameters
     * @param timeoutMs Request timeout in milliseconds (default 30s)
     * @return The result from the response as a [JsonElement]
     * @throws JsonRpcException If the server returns an error
     * @throws kotlinx.coroutines.TimeoutCancellationException If the request times out
     */
    suspend fun request(
        method: String,
        params: JsonElement = JsonObject(emptyMap()),
        timeoutMs: Long = 30_000L
    ): JsonElement {
        val requestId = UUID.randomUUID().toString()
        val deferred = CompletableDeferred<JsonElement>()
        pendingRequests[requestId] = deferred

        val message = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", requestId)
            put("method", method)
            put("params", params)
        }

        sendMessage(message)

        try {
            return withTimeout(timeoutMs) {
                deferred.await()
            }
        } finally {
            pendingRequests.remove(requestId)
        }
    }

    /**
     * Send a JSON-RPC notification (no response expected).
     */
    suspend fun notify(method: String, params: JsonElement = JsonObject(emptyMap())) {
        val message = buildJsonObject {
            put("jsonrpc", "2.0")
            put("method", method)
            put("params", params)
        }
        sendMessage(message)
    }

    // ========================================================================
    // Private Implementation
    // ========================================================================

    private suspend fun sendMessage(message: JsonObject) {
        withContext(Dispatchers.IO) {
            val content = json.encodeToString(JsonObject.serializer(), message)
            val contentBytes = content.toByteArray(StandardCharsets.UTF_8)
            val header = "Content-Length: ${contentBytes.size}\r\n\r\n"

            synchronized(writeLock) {
                outputStream.write(header.toByteArray(StandardCharsets.UTF_8))
                outputStream.write(contentBytes)
                outputStream.flush()
            }
        }
    }

    private fun readLoop() {
        val bis = BufferedInputStream(inputStream)
        try {
            while (running) {
                val message = readMessage(bis) ?: break
                handleMessage(message)
            }
        } catch (e: Exception) {
            if (running) {
                // Connection error - complete all pending with exception
                for ((_, deferred) in pendingRequests) {
                    deferred.completeExceptionally(IOException("Connection lost: ${e.message}"))
                }
            }
        }
    }

    private fun readMessage(input: InputStream): JsonObject? {
        // Read header line character by character
        val headerLine = StringBuilder()
        var c: Int
        while (true) {
            c = input.read()
            if (c == -1) return null
            if (c == '\n'.code) break
            if (c != '\r'.code) headerLine.append(c.toChar())
        }

        val header = headerLine.toString().trim()
        if (!header.startsWith("Content-Length:")) return null

        val contentLength = header.substringAfter("Content-Length:").trim().toInt()

        // Read the empty line separator
        while (true) {
            c = input.read()
            if (c == -1) return null
            if (c == '\n'.code) break
        }

        // Read exact content bytes, handling short reads
        val contentBytes = ByteArray(contentLength)
        var bytesRead = 0
        while (bytesRead < contentLength) {
            val read = input.read(contentBytes, bytesRead, contentLength - bytesRead)
            if (read == -1) throw EOFException("Unexpected end of stream while reading JSON-RPC message")
            bytesRead += read
        }

        val content = String(contentBytes, StandardCharsets.UTF_8)
        return json.parseToJsonElement(content).jsonObject
    }

    private fun handleMessage(message: JsonObject) {
        val id = message["id"]?.let {
            when (it) {
                is JsonPrimitive -> it.content
                else -> it.toString()
            }
        }
        val method = message["method"]?.let {
            (it as? JsonPrimitive)?.content
        }

        // Response to our request (has id + result or error, no method)
        if (id != null && (message.containsKey("result") || message.containsKey("error"))) {
            val deferred = pendingRequests[id]
            if (deferred != null) {
                if (message.containsKey("error")) {
                    val error = message["error"]!!.jsonObject
                    val code = error["code"]?.let { (it as? JsonPrimitive)?.intOrNull } ?: -1
                    val errorMsg = error["message"]?.let { (it as? JsonPrimitive)?.content } ?: "Unknown error"
                    val data = error["data"]
                    deferred.completeExceptionally(JsonRpcException(code, errorMsg, data))
                } else {
                    deferred.complete(message["result"] ?: JsonObject(emptyMap()))
                }
            }
            return
        }

        // Notification from server (has method, no id)
        if (method != null && id == null) {
            val params = message["params"]?.jsonObject ?: JsonObject(emptyMap())
            notificationHandler?.invoke(method, params)
            return
        }

        // Incoming request from server (has both method and id)
        if (method != null && id != null) {
            val params = message["params"]?.jsonObject ?: JsonObject(emptyMap())
            val handler = requestHandlers[method]

            scope.launch {
                if (handler == null) {
                    sendErrorResponse(id, -32601, "Method not found: $method")
                    return@launch
                }
                try {
                    val result = handler(params)
                    sendResponse(id, result)
                } catch (e: JsonRpcException) {
                    sendErrorResponse(id, e.code, e.message ?: "Unknown error")
                } catch (e: Exception) {
                    sendErrorResponse(id, -32603, e.message ?: "Internal error")
                }
            }
        }
    }

    private suspend fun sendResponse(requestId: String, result: JsonObject) {
        val response = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", requestId)
            put("result", result)
        }
        sendMessage(response)
    }

    private suspend fun sendErrorResponse(requestId: String, code: Int, message: String) {
        val response = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", requestId)
            put("error", buildJsonObject {
                put("code", code)
                put("message", message)
            })
        }
        sendMessage(response)
    }
}

File: scala/src/main/scala/com/github/copilot/JsonRpcClient.scala
// Copyright (c) Microsoft Corporation. All rights reserved.

package com.github.copilot

import io.circe.*
import io.circe.parser.*
import io.circe.syntax.*

import java.io.{BufferedInputStream, InputStream, OutputStream}
import java.nio.charset.StandardCharsets
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import scala.concurrent.{ExecutionContext, Future, Promise}
import scala.util.{Failure, Success, Try}

// ============================================================================
// JSON-RPC 2.0 Message Types
// ============================================================================

/** JSON-RPC 2.0 error. */
case class JsonRpcError(code: Int, message: String, data: Option[Json] = None)

object JsonRpcError:
  given Encoder[JsonRpcError] = Encoder.instance { e =>
    Json.obj(
      "code" -> e.code.asJson,
      "message" -> e.message.asJson,
      "data" -> e.data.asJson
    ).dropNullValues
  }
  given Decoder[JsonRpcError] = Decoder.instance { c =>
    for
      code    <- c.get[Int]("code")
      message <- c.get[String]("message")
      data    <- c.get[Option[Json]]("data")
    yield JsonRpcError(code, message, data)
  }

/** Exception wrapping a JSON-RPC error. */
class JsonRpcException(val error: JsonRpcError)
  extends RuntimeException(s"JSON-RPC Error ${error.code}: ${error.message}")

/** JSON-RPC 2.0 request (or notification when id is None). */
private[copilot] case class JsonRpcRequest(
  jsonrpc: String = "2.0",
  id: Option[String],
  method: String,
  params: Json
)

private[copilot] object JsonRpcRequest:
  given Encoder[JsonRpcRequest] = Encoder.instance { r =>
    val base = Json.obj(
      "jsonrpc" -> r.jsonrpc.asJson,
      "method" -> r.method.asJson,
      "params" -> r.params
    )
    r.id match
      case Some(id) => base.deepMerge(Json.obj("id" -> id.asJson))
      case None     => base
  }

/** JSON-RPC 2.0 response. */
private[copilot] case class JsonRpcResponse(
  id: Option[String],
  result: Option[Json],
  error: Option[JsonRpcError]
)

private[copilot] object JsonRpcResponse:
  given Decoder[JsonRpcResponse] = Decoder.instance { c =>
    for
      id     <- c.get[Option[String]]("id")
      result <- c.get[Option[Json]]("result")
      error  <- c.get[Option[JsonRpcError]]("error")
    yield JsonRpcResponse(id, result, error)
  }

/** Incoming message from the server -- either a request/notification or a response. */
private[copilot] case class IncomingMessage(
  id: Option[Json],
  method: Option[String],
  params: Option[Json],
  result: Option[Json],
  error: Option[JsonRpcError]
):
  /** True when this is a request (has method + id). */
  def isRequest: Boolean = method.isDefined && id.isDefined && !id.contains(Json.Null)
  /** True when this is a notification (has method but no id). */
  def isNotification: Boolean = method.isDefined && (id.isEmpty || id.contains(Json.Null))
  /** True when this is a response (has id, no method). */
  def isResponse: Boolean = method.isEmpty && id.isDefined && !id.contains(Json.Null)

  def idString: Option[String] = id.flatMap(_.asString)

private[copilot] object IncomingMessage:
  given Decoder[IncomingMessage] = Decoder.instance { c =>
    for
      id     <- c.get[Option[Json]]("id")
      method <- c.get[Option[String]]("method")
      params <- c.get[Option[Json]]("params")
      result <- c.get[Option[Json]]("result")
      error  <- c.get[Option[JsonRpcError]]("error")
    yield IncomingMessage(id, method, params, result, error)
  }

// ============================================================================
// Request Handler Type
// ============================================================================

/**
 * A handler for incoming server-to-client requests.
 * Returns a Future of the result Json, or fails with a JsonRpcException on error.
 */
type RequestHandler = Json => Future[Json]

// ============================================================================
// JSON-RPC Client
// ============================================================================

/**
 * JSON-RPC 2.0 client using Content-Length header framing.
 *
 * This client communicates over stdio (or any pair of input/output streams) with
 * the Copilot CLI server. Messages are framed using the standard
 * `Content-Length: <n>\r\n\r\n<body>` protocol (same as LSP).
 *
 * @param input   The input stream to read server messages from
 * @param output  The output stream to write client messages to
 * @param ec      The execution context for asynchronous operations
 */
class JsonRpcClient(
  input: InputStream,
  output: OutputStream
)(using ec: ExecutionContext):

  // Pending request promises, keyed by request ID
  private val pendingRequests = new ConcurrentHashMap[String, Promise[Json]]()

  // Registered request handlers (for server-to-client requests)
  private val requestHandlers = new ConcurrentHashMap[String, RequestHandler]()

  // Lock for writing to the output stream
  private val writeLock = new Object()

  @volatile private var running = false
  @volatile private var readThread: Option[Thread] = None

  // -------------------------------------------------------------------------
  // Lifecycle
  // -------------------------------------------------------------------------

  /** Start the background read loop. */
  def start(): Unit =
    if running then return
    running = true
    val thread = new Thread(() => readLoop(), "jsonrpc-read-loop")
    thread.setDaemon(true)
    readThread = Some(thread)
    thread.start()

  /** Stop the client and fail all pending requests. */
  def stop(): Unit =
    running = false
    readThread.foreach(_.interrupt())
    readThread = None
    // Fail all pending requests
    val error = new RuntimeException("JSON-RPC client stopped")
    pendingRequests.forEach { (_, promise) =>
      promise.tryFailure(error)
    }
    pendingRequests.clear()

  // -------------------------------------------------------------------------
  // Outbound: send requests and notifications
  // -------------------------------------------------------------------------

  /**
   * Send a JSON-RPC request and return a Future of the result.
   *
   * @param method  The RPC method name
   * @param params  The parameters as a circe Json value
   * @return        A Future that completes with the result Json
   */
  def sendRequest(method: String, params: Json): Future[Json] =
    val id = UUID.randomUUID().toString
    val promise = Promise[Json]()
    pendingRequests.put(id, promise)

    val request = JsonRpcRequest(id = Some(id), method = method, params = params)
    Try(writeMessage(request.asJson)) match
      case Failure(ex) =>
        pendingRequests.remove(id)
        promise.tryFailure(ex)
      case Success(_) => // message sent, waiting for response

    promise.future

  /**
   * Send a JSON-RPC notification (no response expected).
   *
   * @param method  The RPC method name
   * @param params  The parameters
   */
  def sendNotification(method: String, params: Json): Unit =
    val notification = JsonRpcRequest(id = None, method = method, params = params)
    writeMessage(notification.asJson)

  // -------------------------------------------------------------------------
  // Inbound: register handlers for server-initiated requests
  // -------------------------------------------------------------------------

  /**
   * Register a handler for incoming requests on the given method.
   * The handler receives the `params` Json and must return a Future[Json] result.
   */
  def onRequest(method: String, handler: RequestHandler): Unit =
    requestHandlers.put(method, handler)

  /**
   * Register a handler for incoming notifications on the given method.
   * Notifications are one-way; no response is sent.
   */
  def onNotification(method: String, handler: Json => Unit): Unit =
    requestHandlers.put(method, params =>
      handler(params)
      Future.successful(Json.Null) // notifications don't send responses
    )

  /**
   * Remove a previously registered request/notification handler.
   */
  def removeHandler(method: String): Unit =
    requestHandlers.remove(method)

  // -------------------------------------------------------------------------
  // Wire protocol: Content-Length header framing
  // -------------------------------------------------------------------------

  private def writeMessage(json: Json): Unit =
    val body = json.noSpaces.getBytes(StandardCharsets.UTF_8)
    val header = s"Content-Length: ${body.length}\r\n\r\n"
    writeLock.synchronized:
      output.write(header.getBytes(StandardCharsets.US_ASCII))
      output.write(body)
      output.flush()

  private def readLoop(): Unit =
    // Read from the raw byte stream. Content-Length is a BYTE count, so the body
    // must be read as exactly that many bytes and only then decoded as UTF-8.
    // Reading via a char-based Reader (counting chars) desynchronises framing on
    // any multi-byte UTF-8 payload, bleeding the next message's header into the
    // current body. BufferedInputStream is created once so its buffer persists
    // across messages.
    val in = new BufferedInputStream(input)
    try
      while running do
        // 1. Read headers until blank line
        val contentLength = readHeaders(in)
        if contentLength < 0 then
          // EOF or error
          running = false
        else if contentLength > 0 then
          // 2. Read exactly contentLength bytes
          val body = new Array[Byte](contentLength)
          var totalRead = 0
          var eof = false
          while totalRead < contentLength && !eof do
            val n = in.read(body, totalRead, contentLength - totalRead)
            if n < 0 then eof = true
            else totalRead += n

          if eof then
            running = false
          else
            val bodyStr = new String(body, 0, totalRead, StandardCharsets.UTF_8)
            handleIncomingMessage(bodyStr)
    catch
      case _: InterruptedException => // expected on stop
      case _: java.io.IOException if !running => // stream closed during shutdown
      case ex: Exception =>
        System.err.println(s"[JsonRpcClient] Read loop error: ${ex.getMessage}")
    finally
      running = false

  /**
   * Read HTTP-style headers (byte-wise) and return the Content-Length value.
   * Returns -1 on EOF before the terminating blank line.
   */
  private def readHeaders(in: InputStream): Int =
    var contentLength = -1
    var done = false
    while !done do
      val line = readHeaderLine(in)
      if line == null then return -1 // EOF
      if line.isEmpty then
        done = true // blank line terminates the header block
      else if line.startsWith("Content-Length:") then
        val value = line.substring("Content-Length:".length).trim
        Try(value.toInt).foreach(cl => contentLength = cl)
    contentLength

  /**
   * Read a single CRLF-terminated header line as ASCII, returning it without the
   * trailing CR/LF. Returns null on immediate EOF.
   */
  private def readHeaderLine(in: InputStream): String =
    val sb = new StringBuilder()
    var b = in.read()
    if b < 0 then return null // EOF
    while b >= 0 && b != '\n' do
      if b != '\r' then sb.append(b.toChar)
      b = in.read()
    sb.toString

  // -------------------------------------------------------------------------
  // Message dispatch
  // -------------------------------------------------------------------------

  private def handleIncomingMessage(body: String): Unit =
    parse(body) match
      case Left(err) =>
        System.err.println(s"[JsonRpcClient] Failed to parse message: ${err.getMessage}")
      case Right(json) =>
        json.as[IncomingMessage] match
          case Left(err) =>
            System.err.println(s"[JsonRpcClient] Failed to decode message: ${err.getMessage}")
          case Right(msg) =>
            if msg.isResponse then
              handleResponse(msg)
            else if msg.isRequest then
              handleServerRequest(msg)
            else if msg.isNotification then
              handleServerNotification(msg)

  private def handleResponse(msg: IncomingMessage): Unit =
    msg.idString.foreach { id =>
      val promise = pendingRequests.remove(id)
      if promise != null then
        msg.error match
          case Some(err) => promise.tryFailure(JsonRpcException(err))
          case None      => promise.trySuccess(msg.result.getOrElse(Json.Null))
    }

  private def handleServerRequest(msg: IncomingMessage): Unit =
    val method = msg.method.getOrElse("")
    val params = msg.params.getOrElse(Json.obj())
    val requestId = msg.id.getOrElse(Json.Null)

    val handler = requestHandlers.get(method)
    if handler == null then
      // Method not found -- send error response
      sendErrorResponse(requestId, -32601, s"Method not found: $method")
    else
      // Execute handler asynchronously
      handler(params).onComplete:
        case Success(result) =>
          sendResultResponse(requestId, result)
        case Failure(ex: JsonRpcException) =>
          sendErrorResponse(requestId, ex.error.code, ex.error.message, ex.error.data)
        case Failure(ex) =>
          sendErrorResponse(requestId, -32603, s"Internal error: ${ex.getMessage}")

  private def handleServerNotification(msg: IncomingMessage): Unit =
    val method = msg.method.getOrElse("")
    val params = msg.params.getOrElse(Json.obj())
    val handler = requestHandlers.get(method)
    if handler != null then
      // Fire and forget -- notifications don't get responses
      handler(params).failed.foreach: ex =>
        System.err.println(s"[JsonRpcClient] Notification handler error for '$method': ${ex.getMessage}")

  private def sendResultResponse(id: Json, result: Json): Unit =
    val response = Json.obj(
      "jsonrpc" -> "2.0".asJson,
      "id" -> id,
      "result" -> result
    )
    Try(writeMessage(response)).failed.foreach: ex =>
      System.err.println(s"[JsonRpcClient] Failed to send response: ${ex.getMessage}")

  private def sendErrorResponse(id: Json, code: Int, message: String, data: Option[Json] = None): Unit =
    val errorObj = Json.obj(
      "code" -> code.asJson,
      "message" -> message.asJson,
      "data" -> data.asJson
    ).dropNullValues
    val response = Json.obj(
      "jsonrpc" -> "2.0".asJson,
      "id" -> id,
      "error" -> errorObj
    )
    Try(writeMessage(response)).failed.foreach: ex =>
      System.err.println(s"[JsonRpcClient] Failed to send error response: ${ex.getMessage}")
